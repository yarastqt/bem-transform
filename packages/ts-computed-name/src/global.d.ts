declare module '@bem/sdk.naming.cell.match' {
  interface IEntity {
    toJSON(): {
      block: string,
      elem?: string,
      [key: string]: any,
    },
  }

  function match(conventional: any): (path: string) => {
    cell: {
      entity: IEntity,
    },
    isMatch: boolean,
  }

  export = match
}
