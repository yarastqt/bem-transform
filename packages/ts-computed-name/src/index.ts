import * as TS from 'typescript'
import * as presets from '@bem/sdk.naming.presets'
import match from '@bem/sdk.naming.cell.match'
import pascalCase from 'pascal-case'


const not = (value: boolean) => !value

export interface IOptions {
  availableClassParent?: string[],
  availibleBemName?: boolean,
  availibleDisplayName?: boolean,
  naming?: 'origin' | 'react',
  overrideExistingProperties?: boolean,
}

interface IPropertyOptions {
  name: string,
  value: string,
  isStatic: boolean,
}

interface IEntity {
  block: string,
  elem?: string,
}

function createProperty(options: IPropertyOptions) {
  const modifiers = options.isStatic
    ? [TS.createModifier(TS.SyntaxKind.StaticKeyword)]
    : []

  return TS.createProperty(
    undefined,
    modifiers,
    TS.createIdentifier(options.name),
    undefined,
    undefined,
    TS.createStringLiteral(options.value),
  )
}

function getDisplayName(entity: IEntity) {
  if (entity.elem !== undefined) {
    return `${pascalCase(entity.block)}-${pascalCase(entity.elem)}`
  }

  return pascalCase(entity.block)
}

function hasProperty(members: TS.ClassElement[], names: string[]) {
  return members.some((member) => (
    TS.isPropertyDeclaration(member)
    && TS.isIdentifier(member.name)
    && names.includes(member.name.text)
  ))
}

const defaultConfig = {
  availableClassParent: ['Block', 'Elem'],
  availibleBemName: true,
  availibleDisplayName: true,
  naming: 'origin',
  overrideExistingProperties: true,
}

export function computeClassBemName(config?: IOptions) {
  const options = { ...defaultConfig, ...config }
  // @ts-ignore (Presets is object with keys, but typings has no index signature)
  const parseBemPath = match(presets[options.naming])

  return function transform(context: TS.TransformationContext) {
    return function traverse(sourceFile: TS.SourceFile) {
      function visitor(node: TS.Node): TS.Node {
        if (TS.isClassDeclaration(node)) {
          const { fileName } = sourceFile
          const blockInPath = fileName.match('blocks')

          if (blockInPath === null) {
            return node
          }

          const filePath = fileName.slice(blockInPath.index, fileName.length)
          const { cell, isMatch } = parseBemPath(filePath)

          // Skip this node if then is not BEM entity
          if (not(isMatch)) {
            return node
          }

          // At zero position always should be extend declaration
          // Skip this node if class does not have extended
          if (node.heritageClauses !== undefined && node.heritageClauses[0] !== undefined) {
            const { expression } = node.heritageClauses[0].types[0]
            // Skip this node if class not extend from Block or Elem by default
            if (TS.isIdentifier(expression) && not(options.availableClassParent.includes(expression.text))) {
              return node
            }
          }
          else {
            return node
          }

          // Copy all values because members cannot be rewritten
          let members = [...node.members]
          const entity = cell.entity.toJSON()

          if (options.overrideExistingProperties) {
            const propertyNames = ['block', 'elem']

            if (options.availibleDisplayName) {
              propertyNames.push('displayName')
            }

            members = members.filter((member) => (
              TS.isPropertyDeclaration(member)
              && TS.isIdentifier(member.name)
              && not(propertyNames.includes(member.name.text))
            ))
          }

          if (options.availibleBemName) {
            if (options.overrideExistingProperties || not(hasProperty(members, ['block', 'elem']))) {
              const properties = Object.keys(entity)
                .map((key) => createProperty({
                  name: key,
                  value: entity[key],
                  isStatic: false,
                }))

              members = members.concat(properties)
            }
          }

          if (options.availibleDisplayName) {
            if (options.overrideExistingProperties || not(hasProperty(members, ['displayName']))) {
              const property = createProperty({
                name: 'displayName',
                value: getDisplayName(entity),
                isStatic: true,
              })

              members = members.concat(property)
            }
          }

          return TS.updateClassDeclaration(
            node,
            node.decorators,
            node.modifiers,
            node.name,
            node.typeParameters,
            node.heritageClauses || [],
            members,
          )
        }

        return TS.visitEachChild(node, visitor, context)
      }

      return TS.visitNode(sourceFile, visitor)
    }
  }
}
