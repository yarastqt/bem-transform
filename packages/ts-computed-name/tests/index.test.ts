import { compiler } from '../../../tests/compiler'
import { IOptions, computeClassBemName } from '../src'

// tslint:disable max-line-length
describe('naming::origin', () => {
  const compile = compiler<IOptions>('ts-computed-name/tests/fixtures/origin/blocks', computeClassBemName)

  it('should add block, elem, and displayName property', () => {
    expect(compile('my-block-1/my-block-1.ts')).toMatchSnapshot()
    expect(compile('my-block-1/__elem/my-block-1__elem.ts')).toMatchSnapshot()
  })

  it('should add block, elem property', () => {
    expect(compile('my-block-1/my-block-1.ts', { availibleDisplayName: false })).toMatchSnapshot()
    expect(compile('my-block-1/__elem/my-block-1__elem.ts', { availibleDisplayName: false })).toMatchSnapshot()
  })

  it('should add displayName property', () => {
    expect(compile('my-block-1/my-block-1.ts', { availibleBemName: false })).toMatchSnapshot()
    expect(compile('my-block-1/__elem/my-block-1__elem.ts', { availibleBemName: false })).toMatchSnapshot()
  })

  it('should override block, elem, and displayName property', () => {
    expect(compile('my-block-2/my-block-2.ts')).toMatchSnapshot()
    expect(compile('my-block-2/__elem/my-block-2__elem.ts')).toMatchSnapshot()
  })

  it('should not add block, elem, and displayName property if they already defined', () => {
    expect(compile('my-block-2/my-block-2.ts', { overrideExistingProperties: false })).toMatchSnapshot()
    expect(compile('my-block-2/__elem/my-block-2__elem.ts', { overrideExistingProperties: false })).toMatchSnapshot()
  })

  it('should not add block, elem, and displayName if parent class not incorrect', () => {
    expect(compile('my-block-1/my-block-1.ts', { availableClassParent: ['YBlock', 'YElem'] })).toMatchSnapshot()
    expect(compile('my-block-1/__elem/my-block-1__elem.ts', { availableClassParent: ['YBlock', 'YElem'] })).toMatchSnapshot()
  })
})
