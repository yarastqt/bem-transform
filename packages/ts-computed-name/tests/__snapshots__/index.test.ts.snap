// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`naming::origin should add block, elem property 1`] = `
"class MyBlock extends Block {
    constructor() {
        super(...arguments);
        this.block = \\"my-block-1\\";
    }
}
"
`;

exports[`naming::origin should add block, elem property 2`] = `
"class MyElem extends Elem {
    constructor() {
        super(...arguments);
        this.block = \\"my-block-1\\";
        this.elem = \\"elem\\";
    }
}
"
`;

exports[`naming::origin should add block, elem, and displayName property 1`] = `
"class MyBlock extends Block {
    constructor() {
        super(...arguments);
        this.block = \\"my-block-1\\";
    }
}
MyBlock.displayName = \\"MyBlock_1\\";
"
`;

exports[`naming::origin should add block, elem, and displayName property 2`] = `
"class MyElem extends Elem {
    constructor() {
        super(...arguments);
        this.block = \\"my-block-1\\";
        this.elem = \\"elem\\";
    }
}
MyElem.displayName = \\"MyBlock_1-Elem\\";
"
`;

exports[`naming::origin should add displayName property 1`] = `
"class MyBlock extends Block {
}
MyBlock.displayName = \\"MyBlock_1\\";
"
`;

exports[`naming::origin should add displayName property 2`] = `
"class MyElem extends Elem {
}
MyElem.displayName = \\"MyBlock_1-Elem\\";
"
`;

exports[`naming::origin should not add block, elem, and displayName if parent class not incorrect 1`] = `
"class MyBlock extends Block {
}
"
`;

exports[`naming::origin should not add block, elem, and displayName if parent class not incorrect 2`] = `
"class MyElem extends Elem {
}
"
`;

exports[`naming::origin should not add block, elem, and displayName property if they already defined 1`] = `
"class MyBlock extends Block {
    constructor() {
        super(...arguments);
        this.block = 'Block';
    }
}
MyBlock.displayName = 'Block';
"
`;

exports[`naming::origin should not add block, elem, and displayName property if they already defined 2`] = `
"class MyElem extends Elem {
    constructor() {
        super(...arguments);
        this.block = 'Block';
        this.elem = 'Elem';
    }
}
MyElem.displayName = 'Block';
"
`;

exports[`naming::origin should override block, elem, and displayName property 1`] = `
"class MyBlock extends Block {
    constructor() {
        super(...arguments);
        this.block = \\"my-block-2\\";
    }
}
MyBlock.displayName = \\"MyBlock_2\\";
"
`;

exports[`naming::origin should override block, elem, and displayName property 2`] = `
"class MyElem extends Elem {
    constructor() {
        super(...arguments);
        this.block = \\"my-block-2\\";
        this.elem = \\"elem\\";
    }
}
MyElem.displayName = \\"MyBlock_2-Elem\\";
"
`;
